.github/actions/oci-auth/oidc.py

```diff
@@ -0,0 +1,418 @@
+) -> dict[str, str]:
+    session = requests.Session()
+
+    res = _fetch_with_retries(
+        url=f'{gh_token_url}&audience={oidc_cfg.audience}',
+        session=session,
+        headers={
+            'Authorization': f'Bearer {gh_token}',
+        },
+    )
+    gh_oidc_token = res.json()['value']
+
+    res = _fetch_with_retries(
+        url=(
+            f'https://sts.amazonaws.com/'
+            '?Action=AssumeRoleWithWebIdentity'
+            f'&DurationSeconds={lifetime_seconds}'
+            f'&RoleArn={oidc_cfg.role_to_assume}'
+            f'&RoleSessionName={oidc_cfg.session_name}'
+            f'&WebIdentityToken={gh_oidc_token}'
+            '&Version=2011-06-15'
+        ),
+        session=session,
+        method='POST',
+        headers={
+            'Accept': 'application/json',
+        },
+    )
+    creds = res.json()['AssumeRoleWithWebIdentityResponse']['AssumeRoleWithWebIdentityResult']['Credentials'] # noqa: E501
+    access_key_id = creds['AccessKeyId']
+    secret_access_key = creds['SecretAccessKey']
+    session_token = creds['SessionToken']
+
+    return {
+        'access_key_id': access_key_id,
+        'secret_access_key': secret_access_key,
+        'session_token': session_token,
+    }
+
+
+def authenticate_against_azure(
+    oidc_cfg: AzureOidcConfiguration,
+    gh_token: str,
+    gh_token_url: str,
+    registry: str,
+) -> dict[str, str]:
+    session = requests.Session()
+
+    res = _fetch_with_retries(
+        url=f'{gh_token_url}&audience={oidc_cfg.audience}',
+        session=session,
+        headers={
+            'Authorization': f'Bearer {gh_token}',
+        },
+    )
+    gh_oidc_token = res.json()['value']
+
+    body = {
+        'client_id': oidc_cfg.client_id,
+        'scope': 'https://management.azure.com/.default',
+        'grant_type': 'client_credentials',
+        'client_assertion_type': 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+        'client_assertion': gh_oidc_token,
+    }
+    res = _fetch_with_retries(
+        url=f'https://login.microsoftonline.com/{oidc_cfg.tenant_id}/oauth2/v2.0/token',
+        session=session,
+        method='POST',
+        data=body,
+        headers={
+            'Content-Type': 'application/x-www-form-urlencoded',
+        },
+    )
+    auth_token = res.json()['access_token']
+
+    body = {
+        'grant_type': 'access_token',
+        'service': registry,
+        'access_token': auth_token,
+    }
+    res = _fetch_with_retries(
+        url=f'https://{registry}/oauth2/exchange',
+        session=session,
+        method='POST',
+        data=body,
+        headers={
+            'Content-Type': 'application/x-www-form-urlencoded',
+        },
+    )
+    refresh_token = res.json()['refresh_token']
+
+    username = '00000000-0000-0000-0000-000000000000'
+
+    token = base64.b64encode(f'{username}:{refresh_token}'.encode()).decode()
+
+    return {
+        'auth': token,
+    }
+
+
+def authenticate_against_gar(
```

Comment:
if available, please add links to documentation of hyperscaler specific authentications